### ISSUE TITLE
Building multiple C libraries for the same toolchain

### ISSUE THREAD

**Issue by keith-packard** [crosstool-ng/crosstool-ng#1395]
Embedded applications often use one of three possible C libraries:

 1. newlib
 2. newlib-nano
 3. picolibc

These can be co-installed in the same toolchain and then selected when running the C compiler by using a suitable --specs option. It would be useful if crosstool-ng supported building all three of these libraries so that the selection of which to use could be left to the application.

For a first cut, this could be limited to C applications so that the C++ support in GCC could still target only a single C library.

**Comment by galak** [comment]
For Zephyr we've added support for doing newlib + newlib-nano.  We've effectively just duplicated the options in Kconfig.  Here are references to the changes:

https://github.com/zephyrproject-rtos/crosstool-ng/commit/d52c77e17cd440511d62b4f9f408abb96b9c9572
https://github.com/zephyrproject-rtos/crosstool-ng/commit/a833aaa5df3e8a06761adc1991ef5d518dc147d7
https://github.com/zephyrproject-rtos/crosstool-ng/commit/660490e5d89b89d5450468ba32c26c803f5dd99d

**Comment by galak** [comment]
@cpackham any thoughts on how to approach this?

**Comment by keith-packard** [comment]
This turned out to be quite easy -- just switch picolibc from a core C library to a companion library for the target. GCC's libstdc++ doesn't support picolibc (yet), so there's no need to build picolibc in the multi-stage gcc adventure. #1412 has the changes necessary to make this work.

**Comment by cpackham** [comment]
@galak I don't have any specific ideas on where to start. It certainly seems like a reasonable request. Presumably were really talking only newlib, nano, pico and glibc, musl, uclibc would remain mutually exclusive. 

There is a little bit of an issue w.r.t which features/languages are supported if we allow multiple libcs to be included e.g. from @keith-packard's comment above should we prohibit selecting C++ support if picolibc is one of the selected libcs or if it's just the default or let the user figure it out when they try to compile some C++ code.

**Comment by keith-packard** [comment]
I'm slowly working on patches to libstdc++ to support picolibc; the changes necessary are pretty simple -- replacing use of POSIX I/O functions (read/write/seek) with C functions (stdio). However, until that is ready for use, I think we'll have to accept configurations where newlib and newlib-nano support C++ and picolibc does not. Otherwise, you'll have a toolchain which, upon adding picolibc, suddenly stops working for some users. I don't think it's going to take long to get the GCC patches done, and I think I will switch to doing that work in crosstool-ng as it's going to take a *lot* more effort to get those patches merged upstream.

**Comment by cpackham** [comment]
I don't think this necessarily requires picolibc to support C++ before it can be done but we just need to think about how things should work. It's probably either lowest-common-denominator or superset-of-all-features.

**Comment by keith-packard** [comment]
My theory is that this requirement will only exist until I've managed to sell picolibc to developers currently using newlib or newlib-nano :-)

For zephyr, I think they will want the union of all features instead of the intersection; otherwise adding a C library will reduce the set of applications that can be built with any C library, which seems like the wrong direction.
